consensus/istanbul/proxy/proxy_handler.go

```diff
@@ -0,0 +1,352 @@
+	ph.addProxyPeer = make(chan consensus.Peer)
+	ph.removeProxyPeer = make(chan consensus.Peer)
+
+	ph.proxyHandlerOpCh = make(chan proxyHandlerOpFunc)
+	ph.proxyHandlerOpDoneCh = make(chan struct{})
+
+	ph.sendValEnodeShareMsgsCh = make(chan struct{})
+	ph.newBlockchainEpoch = make(chan struct{})
+
+	ph.proxyHandlerEpochLength = time.Minute
+
+	ph.ps = newProxySet(newConsistentHashingPolicy())
+
+	ph.logger = log.New()
+
+	return ph
+}
+
+// Start begins the proxyHandler event loop
+func (ph *proxyHandler) Start() error {
+	ph.lock.Lock()
+	defer ph.lock.Unlock()
+
+	if ph.running {
+		return ErrStartedProxyHandler
+	}
+	ph.running = true
+
+	ph.loopWG.Add(1)
+	go ph.run()
+
+	log.Info("Proxy handler started")
+	return nil
+}
+
+// Stop stops the goroutine `run` if it is currently running
+func (ph *proxyHandler) Stop() error {
+	ph.lock.Lock()
+	defer ph.lock.Unlock()
+
+	if !ph.running {
+		return ErrStoppedProxyHandler
+	}
+	ph.running = false
+	close(ph.quit)
+	ph.loopWG.Wait()
+
+	log.Info("Proxy handler stopped")
+	return nil
+}
+
+// isRunning returns if `run` is currently running in a goroutine
+func (ph *proxyHandler) isRunning() bool {
+	ph.lock.RLock()
+	defer ph.lock.RUnlock()
+
+	return ph.running
+}
+
+// run handles changes to proxies, validators, and performs occasional check-ins
+// that proxy/validator assignments are as expected
+func (ph *proxyHandler) run() {
+	defer ph.loopWG.Done()
+
+	phEpochTicker := time.NewTicker(ph.proxyHandlerEpochLength)
+	defer phEpochTicker.Stop()
+
+	logger := ph.logger.New("func", "run")
+
+	ph.updateValidators()
+
+loop:
+	for {
+		select {
+		case <-ph.quit:
+			// The proxyHandler was stopped
+			break loop
+
+		case addProxyNodes := <-ph.addProxies:
+			// Got command to add proxy nodes.
+			// Add any unseen proxies to the proxy set and add p2p static connections to them.
+			for _, proxyNode := range addProxyNodes {
+				proxyID := proxyNode.InternalNode.ID()
+				if ph.ps.getProxy(proxyID) != nil {
+					logger.Debug("Proxy is already in the proxy set", "proxyNode", proxyNode, "proxyID", proxyID, "chan", "addProxies")
+					continue
+				}
+				log.Info("Adding proxy node", "proxyNode", proxyNode, "proxyID", proxyID)
+				ph.ps.addProxy(proxyNode)
+				ph.sb.AddPeer(proxyNode.InternalNode, p2p.ProxyPurpose)
+			}
+
+		case rmProxyNodes := <-ph.removeProxies:
+			// Got command to remove proxy nodes.
+			// Remove the proxy and remove the p2p static connection
+			for _, proxyNode := range rmProxyNodes {
+				proxyID := proxyNode.ID()
+				proxy := ph.ps.getProxy(proxyID)
+				if proxy == nil {
+					logger.Warn("Proxy is not in the proxy set", "proxy", proxyNode, "proxyID", proxyID, "chan", "removeProxies")
+					continue
+				}
+
+				logger.Info("Removing proxy node", "proxy", proxy.String(), "chan", "removeProxies")
+
+				if valsReassigned := ph.ps.removeProxy(proxyID); valsReassigned {
+					ph.sendValEnodeShareMsgs()
+					ph.sb.UpdateAnnounceVersion()
```

Comment:
`sendValEnodeShareMsgs` is a `fire and forget` function for the moment; i wonder if there's a problem if while we are sending the `ValEnodeShares` the `UpdateAnnounceVersion` happens.

In go, the idiom is to make functions blocking, and then if we want to call them async just do `go funCall..`. Which i think makes reading code much easier (you don't need to worry about what's happening concurrently). Here for the unaware reader `sendValEnodeShareMsgs()` seem to be blocking when it's not.

Reply:
I made `sendValEnodeShareMsgs` be non blocking.  It actually should be completed by the time that `updateannounceversion` is called (although not totally necessary, since retries from the proxy_handler and the announce_thread will eventually get the correct messages to the right nodes).